packages/foundry/contracts/hooks/NFTLiquidityStaking/NFTLiquidityStakingHook.sol

```diff
@@ -0,0 +1,416 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import { ERC721 } from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { ReentrancyGuard } from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+
+import { IHooks } from "@balancer-labs/v3-interfaces/contracts/vault/IHooks.sol";
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+import { IPoolInfo } from "@balancer-labs/v3-interfaces/contracts/pool-utils/IPoolInfo.sol";
+import { IBasePoolFactory } from "@balancer-labs/v3-interfaces/contracts/vault/IBasePoolFactory.sol";
+import { BaseHooks } from "@balancer-labs/v3-vault/contracts/BaseHooks.sol";
+import { VaultGuard } from "@balancer-labs/v3-vault/contracts/VaultGuard.sol";
+
+import {
+    LiquidityManagement,
+    TokenConfig,
+    PoolSwapParams,
+    HookFlags
+} from "@balancer-labs/v3-interfaces/contracts/vault/VaultTypes.sol";
+
+import { INFTLiquidityStakingHook } from "./Interfaces/INFTLiquidityStakingHook.sol";
+import "./NFTMetadata.sol";
+import "./NFTGovernor.sol";
+import "./RewardToken.sol";
+import "@openzeppelin/contracts/governance/utils/IVotes.sol";
+
+contract NFTLiquidityStakingHook is
+    INFTLiquidityStakingHook,
+    IHooks,
+    BaseHooks,
+    ERC721,
+    Ownable,
+    VaultGuard,
+    ReentrancyGuard,
+    IVotes
+{
+    NFTGovernor public governor;
+    RewardToken public rewardToken;
+    uint256 public constant REWARD_RATE = 1e18;
+
+    mapping(address => mapping(address => StakingInfo)) private _stakingInfoMap;
+    mapping(address => mapping(address => uint256)) public lastUnstakeTime;
+    mapping(uint256 => address) private _tokenIdToPool;
+    uint256 private _tokenIdCounter;
+
+    mapping(uint256 => uint256) public nftTierToFeeDiscount;
+    mapping(uint256 => uint256) public nftTierToVotingPower;
+    mapping(uint256 => uint256) public nftTierToYieldBoost;
+    mapping(uint256 => mapping(address => uint256)) private _votePowerCheckpoints;
+    mapping(uint256 => uint256) private _totalSupplyCheckpoints;
+    uint256 private _currentCheckpoint;
+
+    mapping(address => mapping(uint256 => uint256)) private _ownedTokens;
+    mapping(address => uint256) private _ownedTokensCount;
+
+    uint256 public constant COOLDOWN_PERIOD = 7 days;
+    uint256 public constant UPGRADE_COOLDOWN_PERIOD = 7 days;
+
+    uint256 public constant BRONZE_TIER_THRESHOLD = 30 days;
+    uint256 public constant SILVER_TIER_THRESHOLD = 90 days;
+    uint256 public constant GOLD_TIER_THRESHOLD = 180 days;
+
+    uint256 public constant BRONZE_TIER_AMOUNT = 1000 ether;
+    uint256 public constant SILVER_TIER_AMOUNT = 5000 ether;
+    uint256 public constant GOLD_TIER_AMOUNT = 10000 ether;
```

Comment:
you might consider using an `enum` type to represent the different nft tiers and grouping related data about each tier into a struct

Reply:
> you might consider using an `enum` type to represent the different nft tiers and grouping related data about each tier into a `struct`

Thank you for this excellent suggestion. I've implemented it in the code. We now use an enum NFTTier to represent the different tiers, and a struct TierInfo to group related data about each tier. 

I've made these changes and updated all relevant parts of the contract and test files to ensure consistency with these new implementations. Please let me know if you'd like me to clarify or expand on any of these changes.